Feature/427 decouple signaling and web rtc stuff

Haardt opened the pull request.

Commit:
first attempt for decoupling

Commit:
remove useless comment

Commit:
rename member variable from _mediaConnnection to _connection

Commit:
remove setup and prepareStreams from api

Commit:
remove media connction member variable from ua class

Commit:
delegate dtmf and some refactorings

Commit:
some clean ups

- remove unused methods in mediaconnection ts
- rename sendDtmf to sendInbandDtmf

Commit:
added gitlab configuration

Commit:
remove console.log

Commit:
remove RTCSessionDescription

Commit:
3.7.2

Commit:
clean up and renaming

Commit:
clean up and renaming

jmillan commented:
Apart from the comments, these changes in the interface:

* Replace `receiveSdpOffer` for `createSdpOffer`
* Replace `receiveSdpAnswer` for `createSdpAnswer`

jmillan commented:
@Haardt ,

Let me know when the comments are handled.

Commit:
work on comments

Commit:
Merge branch 'gitlab' into feature/427-Decouple_signaling_and_WebRTC_stuff

Commit:
remove _createRTCConnection function

Commit:
rename functions
- Replace receiveSdpOffer for createSdpOffer
- Replace receiveSdpAnswer for createSdpAnswer

- Replace setRemoteDescription to setRemoteSdp

Haardt commented:
> Let me know when the comments are handled.
Some updates - comment all comments :).

jmillan commented on gulpfile.js:
```diff
@@ -105,7 +105,8 @@ gulp.task('test', function()
     'test/test-UA-no-WebRTC.js',
-    'test/test-digestAuthentication.js'
+    'test/test-digestAuthentication.js',
+    'test/test-mediaConnection.js'
```
this file is missing.

Haardt replied:
Added

jmillan commented on lib/MediaConnection.d.ts:
```diff
@@ -0,0 +1,21 @@
+type MediaConnectionType = 'NodeMediaConnection' | 'BrowserMediaConnection';
+
+export interface MediaConnection extends EventEmitter {
+    receiveSdpOffer(options: any): Promise<RTCSessionDescriptionInit>
```
`createSdpOffer`

`RTCSessionDescriptionInit` is deprecated, use `RTCSessionDescription` instead.

Haardt replied:
Changed options type to RTCOfferOptions and RTCSessionDescriptionInit to RTCSessionDescription

jmillan commented on lib/MediaConnection.d.ts:
```diff
@@ -0,0 +1,21 @@
+export interface MediaConnection extends EventEmitter {
+    receiveSdpOffer(options: any): Promise<RTCSessionDescriptionInit>
+
+    receiveSdpAnswer(options: any): Promise<RTCSessionDescriptionInit>
```
`createSdpAnser`

`RTCSessionDescriptionInit` is deprecated, use `RTCSessionDescription` instead.

Haardt replied:
Changed options type to RTCOfferOptions and RTCSessionDescriptionInit to RTCSessionDescription

jmillan commented on lib/MediaConnection.js:
```diff
@@ -0,0 +1,59 @@
+ * Interface documentation: https://jssip.net/documentation/$last_version/api/mediaconnection/
+ *
+ * interface MediaConnection {
+ *  attribute String example
+ *
```
```suggestion
```

jmillan commented on lib/MediaConnection.js:
```diff
@@ -0,0 +1,59 @@
+        {
+            if (!Utils.isFunction(mediaConnection[method]))
+            {
+                debugerror(`missing or invalid JsSIP.MediaConnector method: ${method}`);
```
```suggestion
                debugerror(`missing or invalid JsSIP.MediaConnection method: ${method}`);
```

Haardt replied:
Changed

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,171 @@
+        return this.rtcPeerConnection.signalingState === 'stable' ? this.rtcPeerConnection.createOffer(options)
+            .then(offer => {
+                return this.setLocalSdp(offer);
+            }) : Promise.resolve();
```
This method is supposed to return a SDP, but here it's returning nothing.

Haardt replied:
Yes, the lib don't use the return value. The same applies to 'createSdpAnswer'.
The lib get the data from the getLocalSdp function.

jmillan replied:
Let's make `createSdpOffer()` return our local SDP always. So we don't unnecessarily mandate to always call `getLocalSdp()` after `createSdpOffer()`.

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,171 @@
+        });
+    }
+
+    setRemoteSdp(type, sdp)
```
Let's use `RTCSessionDescription` as argument please. It's quite a universal object type.

Haardt replied:
The RTCPeerConnection.createOffer use the 'RTCOfferOptions'-Type for the parameter? Should i use 'rtcOfferOptions' as parameter?

jmillan replied:
> Let's use RTCSessionDescription as argument please. It's quite a universal object type.

I was referring to `setRemoteSdp()`

> The RTCPeerConnection.createOffer use the 'RTCOfferOptions'-Type for the parameter? Should i use 'rtcOfferOptions' as parameter?

I guess you are referring to `createSdpOffer()`. If so, yes an optional `options: RTCOfferOptions`  argument would make it.

jmillan commented on lib/MediaConnection.d.ts:
```diff
@@ -0,0 +1,21 @@
+
+    setRemoteSdp(sdp: RTCSessionDescriptionInit): Promise<void>
+
+    sendInbandDTMF(tones: number, duration: number, interToneGap: number): void
```
I'd rename it to `sendDTMF` so we don't expose how DTMFs are sent. This is just an interface.

Haardt replied:
Why i should remove the integrated SIP-DTMF? Actually only the 'inband' generation has to be implemented by the implementation. 
Hm, i can also remove the entire part and include the SIP part in the default implementation

jmillan replied:
Not about SIP at all, but I may implement a `MediaConnection` class that uses whatever technique (does not even need to be standard) to send a DTMF, that's why I suggest renaming it to `sendDTMF`, but nothing to do with SIP.

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,171 @@
+
+    close()
+    {
+        // TODO: close streams
```
```suggestion
```

Haardt replied:
Removed

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    // TODO: used by un/mute functions. How to handle this?
+    getSenders()
```
In order not to expose `RtpSenders` we can just create the method `getLocalTracks()`, and for consistency `getRemoteTracks()`

Haardt replied:
getLocalTracks = getSenders?
getRemoteTracks = getReceivers?

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    // TODO: make private
+    getLocalStreams()
```
Since we are working with tracks, I'd rather have a `getLocalTracks()` method

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    // TODO: make private
+    getRemoteStreams()
```
Same as in previous comment.

jmillan commented on lib/RTCSession.js:
```diff
@@ -64,8 +65,8 @@ module.exports = class RTCSession extends EventEmitter
 
-    // The RTCPeerConnection instance (public attribute).
-    this._connection = null;
+    // The RTCPeerMediaConnection or custom (nodejs) instance (public attribute).
```
```suggestion
    // The MediaConnection instance.
```

jmillan commented on lib/RTCSession.js:
```diff
@@ -293,11 +294,12 @@ module.exports = class RTCSession extends EventEmitter
-    // Check WebRTC support.
-    if (!window.RTCPeerConnection)
+    // Check support.
+    if (!isMediaConnection(mediaConnection))
```
Let's do this check as soon as possible, on constructor.

Haardt replied:
Moved

jmillan commented on lib/RTCSession.js:
```diff
@@ -293,11 +294,12 @@ module.exports = class RTCSession extends EventEmitter
+    if (!isMediaConnection(mediaConnection))
     {
-      throw new Exceptions.NotSupportedError('WebRTC not supported');
+      throw new Exceptions.NotSupportedError('No valid browser or node media connection provided.');
```
```suggestion
      throw new Exceptions.NotSupportedError('No valid MediaConnection instance provided.');
```

Haardt replied:
Changed

jmillan commented on lib/RTCSession.js:
```diff
@@ -377,8 +379,10 @@ module.exports = class RTCSession extends EventEmitter
 
-    // Create a new RTCPeerConnection instance.
-    this._createRTCConnection(pcConfig, rtcConstraints);
+    //TODO: AH Emit the MediaConnection?
```
Since the user is now providing the media connection, there is no need to fire this event.

Haardt replied:
Ok, i removed the connection emits.

jmillan commented on lib/RTCSession.js:
```diff
@@ -504,14 +508,17 @@ module.exports = class RTCSession extends EventEmitter
     const request = this._request;
     const extraHeaders = Utils.cloneArray(options.extraHeaders);
     const mediaConstraints = Utils.cloneObject(options.mediaConstraints);
-    const mediaStream = options.mediaStream || null;
```
We still want to accept `MediaStream`'s, what we are not doing anymore is `getUserMedia`

Haardt replied:
Custom streams can be set with the new mediaconnection interface

jmillan commented on lib/RTCSession.js:
```diff
@@ -711,22 +654,20 @@ module.exports = class RTCSession extends EventEmitter
             this._failed('system', null, JsSIP_C.causes.WEBRTC_ERROR);
 
-            debugerror('emit "peerconnection:setremotedescriptionfailed" [error:%o]', error);
+            debugerror('emit "peerconnection:setremotesdpfailed" [error:%o]', error);
```
Please replace this logs with `mediaconnection:XYZ`

```suggestion
            debugerror('emit "mediaconnection:setremotesdpfailed" [error:%o]', error);
```

Haardt replied:
renamed

Commit:
Rename mediaconnection interface functions

- createSdpOffer
- createSdpAnswer

Commit:
Renaming and small refactorings

Commit:
delete .gitignore

Commit:
added .gitignore

Commit:
removed comments and changed the return type to viod

Commit:
Merge branch 'feature/427-Decouple_signaling_and_WebRTC_stuff' into gitlab

jmillan commented on lib/MediaConnection.js:
```diff
@@ -0,0 +1,57 @@
+ *
+ *  method createSdpOffer(options: RTCOfferOptions): Promise<RTCSessionDescriptionInit>
+ *  method createSdpAnswer(options: RTCOfferOptions): Promise<RTCSessionDescriptionInit>
+ *  method _setLocalSdp(sdp: RTCSessionDescriptionInit): Promise<void>
```
```suggestion
 *  method setLocalSdp(sdp: RTCSessionDescriptionInit): Promise<void>
```

Haardt replied:
Ok, but we leave 
`if (transportType === JsSIP_C.DTMF_TRANSPORT.RFC2833)`
the SIP-DTMF method in the code?

jmillan replied:
yes :+1:

jmillan commented on lib/MediaConnection.d.ts:
```diff
@@ -0,0 +1,21 @@
+
+    createSdpAnswer(options: RTCOfferOptions): Promise<void>
+
+    setLocalSdp(sdp: RTCSessionDescriptionInit): Promise<void>
```
```suggestion
    setLocalSdp(rtcSessionDescription: RTCSessionDescription): Promise<void>
```

Haardt replied:
Renamed

jmillan commented on lib/MediaConnection.d.ts:
```diff
@@ -0,0 +1,21 @@
+
+    setLocalSdp(sdp: RTCSessionDescriptionInit): Promise<void>
+
+    setRemoteSdp(sdp: RTCSessionDescriptionInit): Promise<void>
```
```suggestion
    setRemoteSdp(rtcSessionDescription: RTCSessionDescription): Promise<void>
```

Haardt replied:
Renamed

jmillan commented on lib/MediaConnection.js:
```diff
@@ -0,0 +1,57 @@
+
+debugerror.log = console.warn.bind(console);
+
+/**
```
I would suggest just removing this comments and pointing to the interface definition.

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,155 @@
+
+    constructor(pcConfig, iceConnectionStateCallback)
+    {
+        super();
```
```suggestion
        super();
        
```

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,155 @@
+
+    createSdpAnswer(rtcSessionDescription) {
+
+        return this.rtcPeerConnection.createAnswer(rtcSessionDescription).then(offer => {
```
Same here, `createSdpAnswer()` must return a SDP.

Haardt replied:
I reworked that part a bit.

jmillan commented on lib/MediaConnection.d.ts:
```diff
@@ -0,0 +1,21 @@
+type MediaConnectionType = 'NodeMediaConnection' | 'BrowserMediaConnection';
+
+export interface MediaConnection extends EventEmitter {
+    createSdpOffer(rtcSessionDescription: RTCSessionDescription): Promise<void>
```
```suggestion
    createSdpOffer(options: RTCOfferOptions): Promise<void>
```

Haardt replied:
Renamed

jmillan commented on lib/MediaConnection.d.ts:
```diff
@@ -0,0 +1,21 @@
+export interface MediaConnection extends EventEmitter {
+    createSdpOffer(rtcSessionDescription: RTCSessionDescription): Promise<void>
+
+    createSdpAnswer(options: RTCOfferOptions): Promise<void>
```
```suggestion
    createSdpAnswer(options: RTCAnswerOptions): Promise<void>
```

Haardt replied:
Renamed

jmillan commented on lib/RTCSession.js:
```diff
@@ -1,4 +1,4 @@
-/* globals RTCPeerConnection: false, RTCSessionDescription: false */
+/* globals RTCmediaconnection: false, RTCSessionDescription: false */
```
```suggestion
/* globals RTCSessionDescription: false */
```

jmillan commented on lib/RTCSession.js:
```diff
@@ -1725,7 +1636,7 @@ module.exports = class RTCSession extends EventEmitter
       catch (error)
       {
-        debugerror('close() | error closing the RTCPeerConnection: %o', error);
+        debugerror('close() | error closing the RTCmediaconnection: %o', error);
```
```suggestion
        debugerror('close() | error closing the MediaConnection: %o', error);
```

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,155 @@
+    }
+
+
+    createSdpOffer(rtcSessionDescription) {
```
```suggestion
    createSdpOffer(options) {
```

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,155 @@
+
+    createSdpOffer(rtcSessionDescription) {
+
+        return this.rtcPeerConnection.signalingState === 'stable' ? this.rtcPeerConnection.createOffer(rtcSessionDescription)
```
```suggestion
        return this.rtcPeerConnection.signalingState === 'stable' ? this.rtcPeerConnection.createOffer(options)
```

Commit:
worked on comments

Commit:
remove comments from mediaconnection implementation validation

Commit:
moved method

Commit:
removed instance creation -> do not exists in node

jmillan commented on .gitignore:
```diff
@@ -1,2 +1,5 @@
 /lib-es5/
+.idea
+.github
+dist/@starface/
```
please don't edit this file. You can achieve this by editing your local ~/.gitignore file.

Haardt replied:
I have another private repo and play ping/pong :|. When we are done I decouple the repos and clean them up.

jmillan replied:
We are close to the final review. Let's remove them. 

You can achieve the same by editing your local ~/.gitignore file.

jmillan commented on lib/MediaConnection.d.ts:
```diff
@@ -0,0 +1,23 @@
+import * as EventEmitter from "events";
+
+export interface MediaConnection extends EventEmitter {
+    createSdpOffer(rtcSessionDescription: RTCOfferOptions): Promise<RTCSessionDescription>
```
```suggestion
    createSdpOffer(options: RTCOfferOptions): Promise<RTCSessionDescription>
```

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,158 @@
+
+    createSdpOffer(options) {
+        return this.rtcPeerConnection.createOffer(options).then(offer =>
+            Promise.all([
+                this.setLocalSdp(offer),
+                this._waitForIceReadiness(options),
+                Promise.resolve(this.getLocalSdp())])
+        ).then( result => result[this.LOCAL_DESCRIPTION_INDEX])
```
Can we change this by:

```js
return this.rtcPeerConnection.createOffer(options)
  .then(offer => this.setLocalSdp(offer))
  .then(this._waitForIceCandidates);
```

This must happen one after the other. With `Promise.all()` is gives the impression that things happen concurrently.

Haardt replied:
Yes, sorry - but the sorting must be:
    return this.rtcPeerConnection.createAnswer(options)
      .then((answer) => this.setLocalSdp(answer))
      .then(() => this._waitForIceCandidates(options))
      .then(() => this.getLocalSdp());
?

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,158 @@
+const EventEmitter = require('events').EventEmitter;
+
+module.exports = class RTCPeerMediaConnection extends EventEmitter {
+    get LOCAL_DESCRIPTION_INDEX() { return 2 };
```
Not needed as per the comment below.

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    createSdpAnswer(options) {
+        return this.rtcPeerConnection.createAnswer(options).then(answer =>
+            Promise.all([
+                this.setLocalSdp(answer),
+                this._waitForIceReadiness(options),
+                Promise.resolve(this.getLocalSdp())]))
+            .then(result => result[this.LOCAL_DESCRIPTION_INDEX])
+    }
```
The very same comment as above.

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,158 @@
+        this.rtcPeerConnection.close()
+    }
+
+    /**
+     * RtcPeer specific methods for the opinionated default implementation
+     */
```
```suggestion
    /**
     * Private Methods.
     */
```

jmillan commented on lib/MediaConnection/RTCPeerMediaConnection.js:
```diff
@@ -0,0 +1,158 @@
+        })
+    }
+
+    // Used by exposed peer-connection (tryit-jssip)
```
Since this method is public, move it up in the file, before private methods.

jmillan commented on lib/RTCSession.d.ts:
```diff
@@ -268,7 +272,9 @@ export class RTCSession extends EventEmitter {
   isReadyToReOffer(): boolean;
 
-  answer(options?: AnswerOptions): void;
+  connect(target: string, options: any, initCallback: (RTCSession) => void)
```
```suggestion
  connect(target: string, options: any, initCallback: (RTCSession) => void): void;
```

jmillan commented on lib/UA.d.ts:
```diff
@@ -144,7 +145,7 @@ declare enum UAStatus {
   static get C(): typeof UAStatus;
 
-  constructor(configuration: UAConfiguration);
+  constructor(configuration: UAConfiguration, mediaConnection: MediaConnection);
```
```suggestion
  constructor(configuration: UAConfiguration);
```

jmillan commented on lib/UA.js:
```diff
@@ -44,7 +44,7 @@ module.exports = class UA extends EventEmitter
   }
 
-  constructor(configuration)
+  constructor(configuration, mediaConnection)
```
```suggestion
  constructor(configuration)
```

jmillan commented on package.json:
```diff
@@ -1,8 +1,8 @@
 {
-  "name": "jssip",
```
Please remove any changes from this file.

jmillan commented on lib/RTCSession.js:
```diff
@@ -49,12 +50,16 @@ module.exports = class RTCSession extends EventEmitter
+    if (!isMediaConnection(mediaConnection))
+    {
+      throw new Exceptions.NotSupportedError('No valid MediaConnection instance provided.');
+    }
```
```suggestion
    }
    
```

jmillan commented on lib/RTCSession.js:
```diff
@@ -504,14 +501,17 @@ module.exports = class RTCSession extends EventEmitter
     debug('answer()');
 
+    if (!isMediaConnection(mediaConnection)) {
+      throw new Exceptions.InvalidMediaConnectionError('Invalid MediaConnection implementation!')
```
```suggestion
      throw new Exceptions.InvalidMediaConnectionError('Invalid MediaConnection instance')
```

jmillan replied:
Why not use a `NotSupportedError` here?

jmillan commented on lib/RTCSession.js:
```diff
@@ -2676,31 +2441,11 @@ module.exports = class RTCSession extends EventEmitter
           return;
         }
-
+        console.log("Can't send invite!", error);
```
```suggestion
```

jmillan commented on lib/RTCSession.js:
```diff
@@ -2839,27 +2582,21 @@ module.exports = class RTCSession extends EventEmitter
           {
             // Be ready for 200 with SDP after a 180/183 with SDP.
             // We created a SDP 'answer' for it, so check the current signaling state.
-            if (this._connection.signalingState === 'stable')
```
Why is this condition removed?

jmillan commented:
New reivew done.

@Haardt, can we please rename the files & class names:

* `MediaConnection` -> `MediaConnectionInterface`, since it defines the interface for the media connection
* `RTCPeerMediaConnection` -> `MediaConnection`, since it is an implementation of the interface with that name

I know this does not respect the naming for `Socket`  and `WebSocketImplementation` but we'll fix it afterwards.

Haardt committed:
Update lib/RTCSession.d.ts

Co-authored-by: José Luis Millán <jmillan@aliax.net>

Haardt committed:
Update lib/MediaConnection/RTCPeerMediaConnection.js

Co-authored-by: José Luis Millán <jmillan@aliax.net>

Haardt committed:
Update lib/RTCSession.js

Co-authored-by: José Luis Millán <jmillan@aliax.net>

Haardt committed:
Working on comments

- Execute sequentially (offer/answer)
- Renaming
- Cleanup (removed localStream vars)

Haardt reviewed (commented):
> New reivew done.
> 
> @Haardt, can we please rename the files & class names:
> 
> * `MediaConnection` -> `MediaConnectionInterface`, since it defines the interface for the media connection
> * `RTCPeerMediaConnection` -> `MediaConnection`, since it is an implementation of the interface with that name
> 
> I know this does not respect the naming for `Socket` and `WebSocketImplementation` but we'll fix it afterwards.

Renamed, but in the Java-World the convention is 'Interface', 'InterfaceImpl'. In Typescript: "In general, do not prefix interfaces with I (e.g. IColor). Because the concept of an interface in TypeScript is much more broad than in C# or Java, the IFoo naming convention is not broadly useful" (Handbook)
https://stackoverflow.com/questions/31876947/confused-about-the-interface-and-class-coding-guidelines-for-typescript

Commit:
- renamed the suggested classes
- Removed mediaconnection from JsIP module
- test the mediaconnection implementation

Commit:
Some clean ups and new integration test for the mediaconnection

jmillan commented:
==================================
Delimiter for PR review 29th April 2021
==================================

jmillan commented on .npmrc:
```diff
@@ -1 +0,0 @@
-package-lock=false
```
same, please remove this.

jmillan commented on lib/MediaConnection/MediaConnection.js:
```diff
@@ -0,0 +1,155 @@
+{
+  constructor(pcConfig, iceConnectionStateCallback, rtcPeerConnection = new RTCPeerConnection(pcConfig))
+  {
+    super();
```
```suggestion
    super();
    
```

jmillan commented on lib/MediaConnection/MediaConnection.js:
```diff
@@ -0,0 +1,155 @@
+      return sender.dtmf;
+    };
+
+    getDTMFSender().insertDTMF(tones, duration, interToneGap);
```
This will throw if `getDTMFSender()` returns undefined.

jmillan commented on lib/MediaConnection/MediaConnection.js:
```diff
@@ -0,0 +1,155 @@
+
+module.exports = class MediaConnection extends EventEmitter
+{
+  constructor(pcConfig, iceConnectionStateCallback, rtcPeerConnection = new RTCPeerConnection(pcConfig))
```
No need for `iceConnectionStateCallback`. This class already exposes the `addEventListener()` method, plus it accepts an optional `RTCPeerConnection` instance in construction. So if you are interested in acting upon ice connection state change you can do it directly on the optional `RTCPeerConnection` passed as argument or use the public method of this class `addEventListener()`.

```suggestion
  constructor(pcConfig, rtcPeerConnection = new RTCPeerConnection(pcConfig))
```

jmillan commented on lib/MediaConnectionInterface.d.ts:
```diff
@@ -0,0 +1,23 @@
+import * as EventEmitter from "events";
+
+export interface MediaConnectionInterface extends EventEmitter {
+    createSdpOffer(options: RTCOfferOptions): Promise<RTCSessionDescription>
```
This file is not in sync with the implementation.

Haardt replied:
?

jmillan commented on lib/MediaConnectionInterface.js:
```diff
@@ -0,0 +1,48 @@
+            }
+        });
+    }
+    catch (e)
```
```suggestion
    catch (error)
```

jmillan commented on lib/MediaConnectionInterface.js:
```diff
@@ -0,0 +1,48 @@
+        {
+            if (!Utils.isFunction(mediaConnection[method]))
+            {
+                console.error(method)
```
```suggestion
```

jmillan commented on lib/RTCSession.js:
```diff
@@ -49,12 +50,16 @@ module.exports = class RTCSession extends EventEmitter
 
     super();
 
+    // if (!isMediaConnection(mediaConnection))
+    // {
+    //   throw new Exceptions.NotSupportedError('No valid MediaConnection instance provided.');
+    // }
```
```suggestion
    if (!isMediaConnection(mediaConnection))
    {
      throw new Exceptions.NotSupportedError('No valid MediaConnection instance provided.');
    }
    
```

Haardt replied:
That is not possible. An incoming call (ua.js) calls the constructor without a 'mediaConnection'. The RTCSession.js method 'answer'  save the mediaConnection in the session.

jmillan commented on lib/RTCSession.js:
```diff
@@ -293,12 +293,7 @@ module.exports = class RTCSession extends EventEmitter
-    {
-      throw new Exceptions.NotSupportedError('WebRTC not supported');
-    }
+    const mediaConnection = this._connection;
```
```suggestion
    
    const mediaConnection = this._connection;
```

jmillan commented on lib/RTCSession.js:
```diff
@@ -504,20 +496,22 @@ module.exports = class RTCSession extends EventEmitter
+
     const request = this._request;
     const extraHeaders = Utils.cloneArray(options.extraHeaders);
     const mediaConstraints = Utils.cloneObject(options.mediaConstraints);
```
`mediaConstraints` where used for `getUserMedia()`. Which is not used anymore. Please remove it and all its references.

jmillan commented on lib/RTCSession.js:
```diff
@@ -712,21 +637,19 @@ module.exports = class RTCSession extends EventEmitter
-
         this._connectionPromiseQueue = this._connectionPromiseQueue
-          .then(() => this._connection.setRemoteDescription(offer))
+          .then(() => mediaConnection.setRemoteSdp({type: 'offer', sdp: e.sdp}))
```
We agreed that  `mediaConnection.setRemoteSdp()` would received an argument of type `RTCSessionDescription`.

jmillan commented on lib/RTCSession.js:
```diff
@@ -1036,15 +959,7 @@ module.exports = class RTCSession extends EventEmitter
-        // Insert tones.
-        sender.insertDTMF(tones, duration, interToneGap);
-      }
+      this._connection.sendInbandDTMF(tones, duration, interToneGap);
```
Please, test that this method is working as it was.

```suggestion
      this._connection.sendDTMF(tones, duration, interToneGap);
```

jmillan commented on lib/RTCSession.js:
```diff
@@ -1474,10 +1389,8 @@ module.exports = class RTCSession extends EventEmitter
-
             this._connectionPromiseQueue = this._connectionPromiseQueue
-              .then(() => this._connection.setRemoteDescription(answer))
+              .then(() => this._connection.setRemoteSdp({type: 'answer', sdp: e.sdp}))
```
Must take a `RTCSessionDescription` argument.

jmillan commented on lib/RTCSession.js:
```diff
@@ -1725,7 +1630,7 @@ module.exports = class RTCSession extends EventEmitter
       catch (error)
       {
-        debugerror('close() | error closing the RTCPeerConnection: %o', error);
+        debugerror('close() | error closing the mediaconnection: %o', error);
```
```suggestion
        debugerror('close() | error closing the MediaConnection: %o', error);
```

jmillan commented on lib/RTCSession.js:
```diff
@@ -1876,111 +1754,36 @@ module.exports = class RTCSession extends EventEmitter
-            return Promise.reject(error);
-          });
+        this.emit('sdp', localDescription);
+        this._rtcReady = true;
```
```suggestion
        this._rtcReady = true;
        
```

jmillan commented on lib/RTCSession.js:
```diff
@@ -2321,13 +2122,13 @@ module.exports = class RTCSession extends EventEmitter
         }
 
-        return this._connection.setRemoteDescription(offer)
+        return this._connection.setRemoteSdp({type: 'offer', sdp: e.sdp})
```
Must take a `RTCSessionDescription` as argument.

jmillan commented on lib/RTCSession.js:
```diff
@@ -2676,31 +2434,11 @@ module.exports = class RTCSession extends EventEmitter
           return;
         }
-
+        console.log("Can't send invite!", error);
```
```suggestion
```

jmillan commented on lib/RTCSession.js:
```diff
@@ -2803,16 +2541,14 @@ module.exports = class RTCSession extends EventEmitter
-
         this._connectionPromiseQueue = this._connectionPromiseQueue
-          .then(() => this._connection.setRemoteDescription(answer))
+          .then(() => this._connection.setRemoteSdp(({type: 'answer', sdp: e.sdp})))
```
Must take `RTCSessionDescription` as argument.

jmillan commented on lib/RTCSession.js:
```diff
@@ -2839,45 +2575,39 @@ module.exports = class RTCSession extends EventEmitter
           {
             // Be ready for 200 with SDP after a 180/183 with SDP.
             // We created a SDP 'answer' for it, so check the current signaling state.
-            if (this._connection.signalingState === 'stable')
```
Why is this condition removed?

Haardt replied:
The 'createSdpOffer' checks the state.

jmillan commented on lib/RTCSession.js:
```diff
@@ -2839,45 +2575,39 @@ module.exports = class RTCSession extends EventEmitter
-              {
-                // Handle Session Timers.
-                this._handleSessionTimersInIncomingResponse(response);
+            this._connection.setRemoteSdp({type: 'answer', sdp: e.sdp})
```
Must take a `RTCSessionDescription` as argument.

jmillan commented on lib/RTCSession.js:
```diff
@@ -2992,10 +2722,9 @@ module.exports = class RTCSession extends EventEmitter
-
       this._connectionPromiseQueue = this._connectionPromiseQueue
-        .then(() => this._connection.setRemoteDescription(answer))
+        .then(() => this._mediaConnection.setRemoteSdp(
```
Must take a `RTCSessionDescription` as argument.

jmillan commented on lib/RTCSession.js:
```diff
@@ -3158,10 +2887,8 @@ module.exports = class RTCSession extends EventEmitter
-
         this._connectionPromiseQueue = this._connectionPromiseQueue
-          .then(() => this._connection.setRemoteDescription(answer))
+            .then(() => this._connection.setRemoteSdp({type: 'answer', sdp: e.sdp}))
```
Must take a RTCSessionDescription as argument.

jmillan commented on lib/UA.js:
```diff
@@ -129,6 +130,15 @@ module.exports = class UA extends EventEmitter
   //  High Level API
   // =================
 
+  updateIncomingMediaConnection(mediaConnection) {
+    if (!isMediaConnection(mediaConnection))
+    {
+      throw new Exceptions.NotSupportedError('No valid MediaConnection instance provided.');
+    }
+
+    this._connection = mediaConnection;
+  }
+
```
```suggestion
```

jmillan commented on lib/UA.js:
```diff
@@ -12,6 +12,7 @@ const Parser = require('./Parser');
 const SIPMessage = require('./SIPMessage');
 const sanityCheck = require('./sanityCheck');
 const config = require('./Config');
+const {isMediaConnection} = require("./MediaConnectionInterface");
```
```suggestion
```

jmillan commented on lib/UA.d.ts:
```diff
@@ -115,6 +116,7 @@ export interface UAEventMap {
   newRTCSession: RTCSessionListener;
   newMessage: MessageListener;
   sipEvent: SipEventListener;
+  updateIncomingMediaConnection: void;
```
```suggestion
```

jmillan commented on package.json:
```diff
@@ -1,8 +1,8 @@
 {
-  "name": "jssip",
+  "name": "@starface/jssip",
```
```suggestion
  "name": "jssip",
```

jmillan commented on package.json:
```diff
@@ -31,8 +31,12 @@
+    "jssip-node-websocket": "^3.0.0",
     "sdp-transform": "^2.14.1"
   },
+  "publishConfig": {
+    "@starface:registry": "https://gitlab.com/api/v4/projects/24469951/packages/npm/"
+  },
```
```suggestion
```

jmillan commented on package.json:
```diff
@@ -31,8 +31,12 @@
     "@types/node": "^14.14.13",
     "debug": "^4.3.1",
     "events": "^3.2.0",
+    "jssip-node-websocket": "^3.0.0",
```
```suggestion
```

Haardt replied:
I need this dependency for the tests

jmillan commented on test/include/test-phone.js:
```diff
@@ -0,0 +1,162 @@
+const https = require('https')
```
Let's do HTTP, we are not trying to tests the secure layer.

jmillan commented on test/include/test-phone.js:
```diff
@@ -0,0 +1,162 @@
+module.exports = class TestPhone {
+    webSocketSipServerMock() {
+        const server = https.createServer({
+            cert: "-----BEGIN CERTIFICATE-----\n" +
```
In a JsSIP test you don't want to test WSS, use plain WS instead.

jmillan commented on test/test-mediaConnection.js:
```diff
@@ -0,0 +1,63 @@
+        const server = testPhone.webSocketSipServerMock()
+        const mediaConnection = new TestMediaConnection()
+        const phone1 = await testPhone.registerTestPhone("test1", mediaConnection)
+        const phone2 = await testPhone.registerTestPhone("test2", mediaConnection)
```
Adjusting to JsSIP sytax.

```suggestion
        const phone2 = await testPhone.registerTestPhone("test2", mediaConnection)
        
```

jmillan commented on test/test-mediaConnection.js:
```diff
@@ -0,0 +1,63 @@
+        const phone1 = await testPhone.registerTestPhone("test1", mediaConnection)
+        const phone2 = await testPhone.registerTestPhone("test2", mediaConnection)
+        let offer = '';
+        let answer = '';
```
```suggestion
        let answer = '';
        
```

jmillan commented on test/test-mediaConnection.js:
```diff
@@ -0,0 +1,63 @@
+                setTimeout(() => resolve(), 1000);
+            });
+            await testPhone.call(phone1, "test2", mediaConnection)
+        })
```
```suggestion
        })
        
```

Commit:
Review updates
